File: w3/dz_java/src/Car/Car.java
package Car;

public class Car {
    private double fuelAmount;
    private double fuelCapacity;
    private double fuelConsumption;
    private Point locaction;
    private String model;

    public Car() {
        this.fuelAmount = 0;
        this.fuelCapacity = 60;
        this.fuelConsumption = 0.6;
        this.locaction = new Point(0,0);
        this.model = "Mersedes";
    }

    public Car(double capacity, double consumption, Point location, String model) {
        this.fuelAmount = 0;
        this.fuelCapacity = capacity;
        this.fuelConsumption = consumption;
        this.locaction = location;
        this.model = model;
    }

    public double getFuelAmount() {
        return this.fuelAmount;
    }

    public double getFuelCapacity() {
        return this.fuelCapacity;
    }

    public double getFuelConsumption() {
        return this.fuelConsumption;
    }

    public Point getLocaction() {
        return this.locaction;
    }

    public String getModel() {
        return this.model;
    }

    public void drive(Point destination) throws OutOfFuel {
        double fuelNeeded = this.locaction.distance(destination) * this.fuelConsumption;

        if ( fuelNeeded > this.fuelAmount ) {
            throw new OutOfFuel();
        }
        this.fuelAmount -= fuelNeeded;
        this.locaction = destination;
    }

    public void drive(double x, double y) throws OutOfFuel {
        Point destination = new Point(x, y);
        this.drive(destination);
    }

    public void refill(double fuel) throws ToMuchFuel {
        if ( this.fuelAmount + fuel > this.fuelCapacity ) {
            throw new ToMuchFuel();
        }
        this.fuelAmount += fuel;
    }

    @Override
    public String toString() {
        return String.format("Model: " + this.getModel() + "\n"
                            + "Consumption: " + this.getFuelConsumption() +"\n"
                            + "Fuel: " + this.getFuelAmount() + "/" + this.getFuelCapacity() + "\n"
                            + "Location: " + this.getLocaction());
    }
}

File: w3/dz_cpp/bcw4/bcw4/Car/Car.cpp
 #include <iostream>
#include "Car.h"
#include "Point.h"

Car::Car(double capacity, double consumption, const Point& location, const std::string& model) {
    this->fuelAmount = 0;
    this->fuelCapacity = capacity;
    this->fuelConsumption = consumption;
    this->location = location;
    this->model = model;
}

Car::~Car() {}

double Car::getFuelAmount() const {
    return this->fuelAmount;
}

double Car::getFuelCapacity() const {
    return this->fuelCapacity;
}

double Car::getFuelConsumption() const {
    return this->fuelConsumption;
}

const Point& Car::getLocation() const {
    return this->location;
}

const std::string& Car::getModel() const {
    return this->model;
}

void Car::drive(const Point& destination) {
    int fuelNeeded = this->location.distance(destination) * this->fuelConsumption;

    if ( fuelNeeded > this->fuelAmount ) {
        throw OutOfFuel();
    }
    this->fuelAmount -= fuelNeeded;
    this->location = destination;
}

void Car::drive(double x, double y) {
    Point destination(x, y);
    this->drive(destination);
}

void Car::refill(double fuel) {
    if ( this->fuelAmount + fuel > this->fuelCapacity ) {
        throw ToMuchFuel();
    }
    this->fuelAmount += fuel;
}

std::ostream& operator<<(std::ostream& out, const Car& car) {
    out << "Car model: " << car.getModel() << "\n";
    out << "Consumption: " << car.getFuelConsumption() << "\n";
    out << "Fuel: " << car.getFuelAmount() << "/" << car.getFuelCapacity() << "\n";
    out << "Location: " << car.getLocation();
    return out;
}